drivers/nutdrv_qx.c

```diff
@@ -1838,20 +1842,45 @@ static int	armac_command(const char *cmd, char *buf, size_t buflen)
 			tmpbuf[0], tmpbuf[1], tmpbuf[2], tmpbuf[3], tmpbuf[4], tmpbuf[5],
 			tmpbuf[1], tmpbuf[2], tmpbuf[3], tmpbuf[4], tmpbuf[5]);
 
+		/*
+		 * On most tested devices (including R/2000I/PSW) this was equal to the number of
+		 * bytes returned in the buffer, but on some newer UPS (R/3000I/PF1) it was 1 more
+		 * (1 control + 5 bytes transferred and bytes_available equal to 6 instead of 5).
+		 *
+		 * Current assumption is that this is number of bytes available on the UPS side
+		 * with up to 5 (ret - 1) transferred.
+		 */
 		bytes_available = (unsigned char)tmpbuf[0] & 0x0f;
 		if (bytes_available == 0) {
 			/* End of transfer */
 			break;
 		}
 
-		memcpy(buf + bufpos, tmpbuf + 1, bytes_available);
-		bufpos += bytes_available;
+		if (bytes_available > READ_SIZE - 1) {
+			/* Single interrupt transfer has 1 control + 5 data bytes */
+			bytes_available = READ_SIZE - 1;
+		}
+
+		/* Copy bytes into the final buffer while detecting end of line - \r */
+		for (size_t i = 0; i < bytes_available; i++) {
+			buf[bufpos++] = tmpbuf[i + 1];
+			if (tmpbuf[i + 1] == 0x0d) {
+				if (i + 1 != bytes_available) {
+					upsdebugx(3, "trailing bytes in serial transmission found: %lu copied out of %lu",
```

Comment:
`%" PRIuSIZE` not `%lu` technically to be clean, not without casting ;)

May need `nut_stdint.h` if not there yet (unlikely).

Reply:
Fixed; I tried %z even before but that didn't really work.